src/main/java/seedu/address/ui/MainWindow.java

```diff
@@ -176,6 +176,15 @@ private CommandResult executeCommand(String commandText) throws CommandException
                 handleSwitchView(commandResult.getTargetView().trim());
             }
 
+            if (commandResult.isUndo()) {
+                logic.getModel().setState((ElisaStateManager)
+                        logic.getModel().getElisaStateHistory().peekCommand().deepCopy());
+            } else {
+                logic.getModel().getElisaStateHistory().pushCommand(logic.getModel().getState().deepCopy());
```

Comment:
This chain of command seems to be a bit long, and it seems to be acting on the model largely, so it might be good to implement a method in model which could be called to do this. In that way we will not need to expose the "stranger" to logic.